### PR TITLE
Add signature retry cache for idempotent durable tasks

### LINKED ISSUE #53: Implement swarm idempotency for task retries
## Summary

When tasks fail and are retried, the current implementation does not handle idempotency properly for swarm and chain patterns. This can lead to orphaned swarms, duplicate task executions, and data inconsistency issues.

### Swarm Retry Issues
- When a task fails mid-swarm creation, many child tasks may already be published
- On retry, a new swarm is created while the old one continues executing
- The original swarm never gets closed/deleted (until TTL expires)
- This leads to orphaned swarms and wasted resources

### Chain Retry Issues  
- When a chain is created and signatures are distributed, tasks start executing
- If the original task crashes mid-run, the first task becomes part of an incomplete workflow
- On retry, the new task may try to access data that was already deleted by the first task's execution
- This causes both the incomplete original chain and the new chain to fail

## Questions to Investigate

- [ ] Can we leverage Hatchet's task caching mechanism for already-published tasks?
- [ ] Is there a way to query task publication status from Hatchet?
- [ ] Can we extract idempotency data from Hatchet's internal state?
- [ ] Should we implement our own caching layer for published tasks?
- [ ] How do we design this to be portable for other task managers (e.g., TaskIQ)?

## Implementation Considerations

1. **Idempotency Key Generation**
   - Need a deterministic way to identify retry attempts vs new executions
   - Keys should incorporate task parameters and workflow context

2. **State Tracking**
   - Track which signatures have been published
   - Store swarm/chain metadata for recovery

3. **Recovery Mechanism**
   - On retry, detect existing swarm/chain state
   - Resume or cleanup based on current status

4. **Task Manager Abstraction**
   - Design should work across different backends
   - Consider a pluggable caching interface

## Tasks

- [ ] Research Hatchet's idempotency capabilities and caching behavior
- [ ] Design idempotency key generation strategy
- [ ] Implement task publication tracking
- [ ] Add swarm/chain recovery logic
- [ ] Create abstraction layer for multi-backend support
- [ ] Add integration tests for retry scenarios

### PR DESCRIPTION
**Pull request by yedidyakfir**
## Summary

Implements a Redis-backed retry cache that ensures signature idempotency when durable tasks are retried. When a task fails and Hatchet retries it, the same signatures (task, chain, swarm) are returned from cache instead of creating duplicates.

## Changes

- Added `SignatureRetryCache` model (Redis-backed with 24h TTL) and `RetryCacheState` for managing cache lifecycle via context variables
- Integrated cache into `sign()`, `chain()`, and `swarm()` creators — on first attempt signatures are cached, on retry they are retrieved from cache in order
- Updated `handle_task_callback` with `is_idempotent` flag — automatically enabled for `durable_task`, manages cache setup/teardown with proper contextvar cleanup
- Added ruff lint config for `third-magic`
- Added comprehensive unit tests for both `third-magic` (creator-level caching) and `mageflow` (callback integration)
- Added integration test verifying signature keys match across retry attempts

## Testing

- Unit tests: `libs/third-magic/tests/unit/creation/test_retry_cache.py` (18 tests)
- Unit tests: `libs/mageflow/tests/unit/callbacks/test_retry_cache.py` (10 tests)
- Integration test: `libs/mageflow/tests/integration/hatchet/test_retry_cache.py`

Closes #53

## Summary by CodeRabbit

* **New Features**
  * Durable tasks gain idempotent retry caching (Redis-backed, 24h) to prevent duplicate signatures on retries.
  * New awaitable task run path to wait for workflow results.
  * Convenience: new remove helper to delete signatures by key.

* **Behavior Changes**
  * Increased default removed-item TTL for longer retention.

* **Tests**
  * Extensive unit and integration tests covering idempotency, cache reuse, cleanup, and edge cases.

* **Documentation**
  * New idempotency guide and changelog entry.

### CODE REVIEW

File: libs/third-magic/pyproject.toml [hunk 1]
```diff
@@ -34,6 +34,9 @@ markers = [
     "hatchet: tests that require hatchet SDK fixtures",
 ]
 
+[tool.ruff.lint]
+select = ["F401", "I"]
```
[the remark this answers is not in the document]
Comment (yedidyakfir) [reply libs/third-magic/pyproject.toml]: I want to explicitly chose my format

File: libs/third-magic/thirdmagic/signature/retry_cache.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    signature_ids: RedisList[RapyerKey] = Field(default_factory=list)
+
+    Meta: ClassVar[RedisConfig] = RedisConfig(ttl=24 * 60 * 60, refresh_ttl=False)
+
+
+@dataclass
+class RetryCacheState:
+    workflow_id: str
+    is_retry: bool
+    cache: SignatureRetryCache
```
[the remark this answers is not in the document]
Comment (yedidyakfir) [reply libs/third-magic/thirdmagic/signature/retry_cache.py]: This is solved when we check  if cache is None: - we recreate the cache

File: libs/third-magic/thirdmagic/task/creator.py [hunk 3]
```diff
@@ -82,6 +90,10 @@ async def sign(task: str | HatchetTaskType, **options: Any) -> TaskSignature:
...
     if isinstance(task, str):
-        return await TaskSignature.from_task_name(task, kwargs=options, **kwargs)
+        signature = await TaskSignature.from_task_name(task, kwargs=options, **kwargs)
     else:
-        return await TaskSignature.from_task(task, kwargs=options, **kwargs)
+        signature = await TaskSignature.from_task(task, kwargs=options, **kwargs)
+
+    await cache_signature(signature)
+
+    return signature
```
[the remark this answers is not in the document]
Comment (yedidyakfir) [reply libs/third-magic/thirdmagic/task/creator.py]: We save signature and cache in transaction now
